velox/exec/tests/TableWriteTest.cpp

```diff
@@ -610,3 +644,96 @@ TEST_F(TableWriteTest, writeNoFile) {
   execute(plan, std::make_shared<core::QueryCtx>(executor_.get()));
   ASSERT_TRUE(fs::is_empty(outputDirectory->path));
 }
+
+TEST_F(TableWriteTest, immutablePartitionsExistingTableError) {
+  auto input = makeVectors(rowType_, 10, 10);
+  auto outputDirectory = TempDirectoryPath::create();
+  auto plan = createInsertPlan(
+      input,
+      rowType_,
+      connector::hive::LocationHandle::TableType::kExisting,
+      outputDirectory->path);
+
+  VELOX_ASSERT_THROW(
+      AssertQueryBuilder(plan)
+          .connectorConfig(
+              kHiveConnectorId,
+              HiveConfig::kImmutablePartitions,
+              folly::to<std::string>("true"))
+          .copyResults(pool()),
+      "Unpartitioned Hive tables are immutable.");
+}
+
+TEST_F(TableWriteTest, immutablePartitionsNewTable) {
+  // When table type is NEW, we always return UpdateMode::kNew, no exception
+  // should be thrown.
+  auto input = makeVectors(rowType_, 10, 10);
+  auto outputDirectory = TempDirectoryPath::create();
+  auto plan = createInsertPlan(
+      input,
+      rowType_,
+      connector::hive::LocationHandle::TableType::kNew,
+      outputDirectory->path);
+
+  auto result = AssertQueryBuilder(plan)
+                    .connectorConfig(
+                        kHiveConnectorId,
+                        HiveConfig::kImmutablePartitions,
+                        folly::to<std::string>("true"))
+                    .copyResults(pool());
+
+  assertEqualResults(
+      {makeRowVector({makeConstant<int64_t>(100, 1)})}, {result});
+}
+
+TEST_F(TableWriteTest, immutablePartitionsInsertRows) {
+  for (auto numRows : {0, 10, 100}) {
+    auto input = makeVectors(rowType_, 10, numRows);
+
+    createDuckDbTable(input);
+
+    for (auto tableType : tableTypes) {
+      auto outputDirectory = TempDirectoryPath::create();
+      auto plan =
+          createInsertPlan(input, rowType_, tableType, outputDirectory->path);
+      assertQuery(plan, "SELECT count(*) from tmp");
+      assertQuery(
+          PlanBuilder().tableScan(rowType_).planNode(),
+          makeHiveConnectorSplits(outputDirectory),
+          "SELECT * FROM tmp");
+    }
+  }
+}
+
+TEST_F(TableWriteTest, immutablePartitionsInsertRowsSequence) {
```

Comment:
This test name is a bit confusing since "immutable_partitions" is false. Maybe insertRowsSequenceUnpartitionedTable is a better test name. Also please explicitly state that this test is for a default immutablePartitions value of false allowing writing into such tables.